contracts/0.8.9/LidoOracleNew.sol

```diff
@@ -0,0 +1,571 @@
+    );
+
+    event ContractVersionSet(uint256 version);
+
+
+    struct MemberReport {
+        // Consensus info
+        uint256 epochId;
+        // CL values
+        uint256 beaconValidators;
+        uint64 beaconBalanceGwei;
+        address[] stakingModules;
+        uint256[] nodeOperatorsWithExitedValidators;
+        uint64[] exitedValidatorsNumbers;
+        // EL values
+        uint256 wcBufferedEther;
+        // decision
+        uint256 newDepositBufferWithdrawalsReserve;
+        uint256[] requestIdToFinalizeUpTo;
+        uint256[] finalizationPooledEtherAmount;
+        uint256[] finalizationSharesAmount;
+    }
+
+    /// ACL
+    bytes32 constant public MANAGE_MEMBERS_ROLE = keccak256("MANAGE_MEMBERS_ROLE");
+    bytes32 constant public MANAGE_QUORUM_ROLE = keccak256("MANAGE_QUORUM_ROLE");
+    bytes32 constant public SET_BEACON_SPEC_ROLE = keccak256("SET_BEACON_SPEC_ROLE");
+    bytes32 constant public SET_REPORT_BOUNDARIES_ROLE = keccak256("SET_REPORT_BOUNDARIES_ROLE");
+    bytes32 constant public SET_BEACON_REPORT_RECEIVER_ROLE = keccak256("SET_BEACON_REPORT_RECEIVER_ROLE");
+
+    /// Eth1 denomination is 18 digits, while Eth2 has 9 digits. Because we work with Eth2
+    /// balances and to support old interfaces expecting eth1 format, we multiply by this
+    /// coefficient.
+    uint128 internal constant DENOMINATION_OFFSET = 1e9;
+
+    /// Historic data about 2 last completed reports and their times
+    bytes32 internal constant POST_COMPLETED_TOTAL_POOLED_ETHER_POSITION = keccak256("lido.LidoOracle.postCompletedTotalPooledEther");
+    bytes32 internal constant PRE_COMPLETED_TOTAL_POOLED_ETHER_POSITION = keccak256("lido.LidoOracle.preCompletedTotalPooledEther");
+    bytes32 internal constant LAST_COMPLETED_EPOCH_ID_POSITION = keccak256("lido.LidoOracle.lastCompletedEpochId");
+    bytes32 internal constant TIME_ELAPSED_POSITION = keccak256("lido.LidoOracle.timeElapsed");
+
+    /// Address of the Lido contract
+    bytes32 internal constant LIDO_POSITION = keccak256("lido.LidoOracle.lido");
+
+    /// Version of the initialized contract data
+    /// NB: Contract versioning starts from 1.
+    /// The version stored in CONTRACT_VERSION_POSITION equals to
+    /// - 0 right after deployment when no initializer is invoked yet
+    /// - N after calling initialize() during deployment from scratch, where N is the current contract version
+    /// - N after upgrading contract from the previous version (after calling finalize_vN())
+    bytes32 internal constant CONTRACT_VERSION_POSITION = keccak256("lido.LidoOracle.contractVersion");
+
+    /// Receiver address to be called when the report is pushed to Lido
+    bytes32 internal constant BEACON_REPORT_RECEIVER_POSITION = keccak256("lido.LidoOracle.beaconReportReceiver");
+
+    /// Upper bound of the reported balance possible increase in APR, controlled by the governance
+    bytes32 internal constant ALLOWED_BEACON_BALANCE_ANNUAL_RELATIVE_INCREASE_POSITION =
```

Comment:
will be replaced
see: https://hackmd.io/R6SKBwd7Tke0d7yg8R_mwg#Safety-nets
and https://hackmd.io/IFhs6ZhRQ0uUFKDVmN17BQ#Beacon-balance-splitting